### PR TITLE
fully answered qustions and features

### PR DESCRIPTION
**Pull request by matangino7**

### CODE REVIEW

File: drf/open_questions.txt [hunk 1]
```diff
@@ -0,0 +1,70 @@
+Request:
+2. we can do it with request.data
+3. if you want to pass parameters with the url https://example.com/resource?param1=value1&param2=value2
```
Comment (omega6team) [inline drf/open_questions.txt:3]: you didn't explain why we want to pass parameters in the url

File: drf/book/models.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+from django.db import models
+from rest_framework import serializers
+from datetime import datetime as dt
+
+
+def validate_publish_date(self, value):
```
Comment (omega6team) [inline drf/book/models.py:6]: no need for self, creates errors

File: drf/open_questions.txt [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+Serializers:
+7. the save() function in Django determines whether an object is new or existing based on the primary key value, and it handles the database operations accordingly. If the primary key is set (not None), it updates the existing record; otherwise, it creates a new one.
+8. errors = serializer.errors
+    print(errors)
+
+9. raise_exception=True is a way drf raise an exception if the data is not valid it will raise a serializers.ValidationError exception if the data is not valid.
+if not serializer.is_valid():
+        raise exceptions.ValidationError(serializer.errors)
+
+11. i will need to create a serializer for that one field and update the view accordingly.
```
Comment (omega6team) [inline drf/open_questions.txt:29]: you didn't add the view class to an url

File: drf/book/serializer.py
Comment (omega6team) [inline drf/book/serializer.py:1]: didn't do question 13

File: drf/open_questions.txt [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+12. depth is the amount of levels you want to serialize inside the object, if i want the author data to be serialized and available to you i will set depth=1.
+
+14. Read-only fields are included in the API output, but should not be included in the input during create or update operations. Any 'read_only' fields that are incorrectly included in the serializer input will be ignored.
```
Comment (omega6team) [inline drf/open_questions.txt:33]: didn't implement it in code

File: drf/open_questions.txt [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+15. SerializerMethodField is a versatile field that allows you to include custom data or calculated values in your serialized data. It is a read-only field, meaning that it can be used to display additional information that doesn't directly correspond to a model field
```
Comment (omega6team) [inline drf/open_questions.txt:35]: didn't implement it in code

File: drf/open_questions.txt [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+Serializer Relations:
+
+16. In drf, a SlugRelatedField acts as a translator for related objects, enabling the representation of these objects using user-friendly slugs instead of numerical primary keys.
+if the read_only=False we need to make sure that unique=True
```
Comment (omega6team) [inline drf/open_questions.txt:40]: didn't write an exapmle

File: drf/open_questions.txt [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+17. select related is essentially a way to query sql join so we can get all the related objects data as well, pre-fetched related will get all related objects using more than one query that will be more useful in a many to many relationship,
```
Comment (omega6team) [inline drf/open_questions.txt:42]: didn't write an exapmle

File: drf/open_questions.txt [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        fields = ('id', 'content', 'timestamp', 'sender')
+
+and than in the user serializer we will create a field that will call the Message serializer:
+
+from rest_framework import serializers
+from .models import User
+from .serializers import MessageSerializer
+
+class UserSerializer(serializers.ModelSerializer):
+    messages = MessageSerializer(many=True, read_only=True)
```
Comment (omega6team) [inline drf/open_questions.txt:61]: you could use what we just learned- SlugRelatedField, to show the messages

File: drf/book/views.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    
+
+class RetrieveUpdateDestroyBook(APIView):
+    """
+    View to retrieve, update, or delete a specific book.
+
+    * Requires token authentication.
+    * Only admin users are able to access this view.
+    """
+    def get(self, request, pk, format=None):
```
Comment (omega6team) [inline drf/book/views.py:57]: the method doesn't return the data of the book, it returns: "idtitleauthorpublication_date"

File: drf/book/views.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        """
+        Delete a specific book by primary key.
+        """
+        try:
+            book = Book.objects.get(pk=pk)
+        except Book.DoesNotExist:
+            return JsonResponse({"error": "Book not found"}, status=404)
+
+        book.delete()
+        return JsonResponse(status=200)
```
Comment (omega6team) [inline drf/book/views.py:94]: error in return: we didn't add data

File: drf/book/models.py (hunk 2 above)
Comment (matangino7) [reply drf/book/models.py:6]: this is how documentation said to do it

File: drf/book/serializer.py
Comment (matangino7) [reply drf/book/serializer.py:1]: deleted on commit, if you want i can send it again

File: drf/open_questions.txt (hunk 4 above)
Comment (matangino7) [reply drf/open_questions.txt:33]: the question didnt saId to implement in code

File: drf/open_questions.txt (hunk 8 above)
Comment (matangino7) [reply drf/open_questions.txt:61]: ok,  but question didn't clarify to use it, i will use slugrelatedfield next time.

**Review by omega6team — APPROVED** [review]

**Pull request merged**